cpp/src/ffi_go/ir/serializer.cpp

```diff
@@ -0,0 +1,153 @@
+
+#include <ffi_go/defs.h>
+#include <ffi_go/ir/LogTypes.hpp>
+#include <ffi_go/LogTypes.hpp>
+
+namespace ffi_go::ir {
+using namespace ffi;
+using namespace ffi::ir_stream;
+
+namespace {
+    /**
+     * Generic helper for ir_serializer_serialize_*_log_event
+     */
+    template <class encoded_variable_t>
+    auto serialize_log_event(
+            StringView log_message,
+            epoch_time_ms_t timestamp_or_delta,
+            void* ir_serializer,
+            ByteSpan* ir_view
+    ) -> int {
+        Serializer* serializer{static_cast<Serializer*>(ir_serializer)};
+        serializer->m_ir_buf.clear();
+
+        bool success{false};
+        if constexpr (std::is_same_v<encoded_variable_t, eight_byte_encoded_variable_t>) {
+            success = eight_byte_encoding::encode_message(
+                    timestamp_or_delta,
+                    std::string_view{log_message.m_data, log_message.m_size},
+                    serializer->m_logtype,
+                    serializer->m_ir_buf
+            );
+        } else if constexpr (std::is_same_v<encoded_variable_t, four_byte_encoded_variable_t>) {
+            success = four_byte_encoding::encode_message(
+                    timestamp_or_delta,
+                    std::string_view{log_message.m_data, log_message.m_size},
+                    serializer->m_logtype,
+                    serializer->m_ir_buf
+            );
+        } else {
+            static_assert(cAlwaysFalse<encoded_variable_t>, "Invalid/unhandled encoding type");
+        }
+        if (false == success) {
+            return static_cast<int>(IRErrorCode_Corrupted_IR);
+        }
+
+        ir_view->m_data = serializer->m_ir_buf.data();
+        ir_view->m_size = serializer->m_ir_buf.size();
+        return static_cast<int>(IRErrorCode_Success);
+    }
+}  // namespace
+
+extern "C" auto ir_serializer_close(void* ir_serializer) -> void {
+    // NOLINTNEXTLINE(cppcoreguidelines-owning-memory)
+    delete static_cast<Serializer*>(ir_serializer);
+}
+
+extern "C" auto ir_serializer_serialize_eight_byte_preamble(
+        StringView ts_pattern,
+        StringView ts_pattern_syntax,
+        StringView time_zone_id,
+        void** ir_serializer_ptr,
+        ByteSpan* ir_view
+) -> int {
+    Serializer* serializer{new Serializer{}};
+    *ir_serializer_ptr = serializer;
+    if (false
+        == eight_byte_encoding::encode_preamble(
+                std::string_view{ts_pattern.m_data, ts_pattern.m_size},
+                std::string_view{ts_pattern_syntax.m_data, ts_pattern_syntax.m_size},
+                std::string_view{time_zone_id.m_data, time_zone_id.m_size},
+                serializer->m_ir_buf
+        ))
+    {
+        return static_cast<int>(IRErrorCode_Corrupted_IR);
+    }
+
+    ir_view->m_data = serializer->m_ir_buf.data();
+    ir_view->m_size = serializer->m_ir_buf.size();
+    return static_cast<int>(IRErrorCode_Success);
+}
+
+extern "C" auto ir_serializer_serialize_four_byte_preamble(
```

Comment:
The function has a lot of duplications with `ir_serializer_serialize_eight_byte_preamble`.
Shall we make a template class instead?

Reply:
Nvm done.